qiskit/qpy/__init__.py

```diff
@@ -105,9 +105,203 @@
+- ``s``: :class:`~.AlignSequential`
+- ``e``: :class:`~.AlignEquispaced`
+
+Note that :class:`~.AlignFunc` context is not supported becasue of the callback function
+stored in the context parameters.
+
+This alignment block is further followed by ``num_element`` length of block elements which may
+consist of nested schedule blocks and schedule instructions.
+Each schedule instruction starts with ``char`` representing the instruction type
+followed by the :ref:`qpy_sequence` block representing the instruction
+:attr:`~qiskit.pulse.instructions.Instruction.operands`.
+Note that the data structure of pulse :class:`~qiskit.pulse.instructions.Instruction`
+is unified so that instance can be uniquely determied by the class and a tuple of operands.
+The mapping of type char to the instruction subclass is defined as follows:
+
+- ``a``: :class:`~qiskit.pulse.instructions.Acquire` instruction
+- ``p``: :class:`~qiskit.pulse.instructions.Play` instruction
+- ``d``: :class:`~qiskit.pulse.instructions.Delay` instruction
+- ``f``: :class:`~qiskit.pulse.instructions.SetFrequency` instruction
+- ``g``: :class:`~qiskit.pulse.instructions.ShiftFrequency` instruction
+- ``q``: :class:`~qiskit.pulse.instructions.SetPhase` instruction
+- ``r``: :class:`~qiskit.pulse.instructions.ShiftPhase` instruction
+- ``b``: :class:`~qiskit.pulse.instructions.RelativeBarrier` instruction
+
+The operands of these instances can be serialized through the standard QPY value serialization
+mechanism, however there are special object types that only appear in the schedule operands.
+Since the operands are serialized as :ref:`qpy_sequence`, each element must be packed with the
+INSTRUCTION_PARAM pack struct, where each payload starts with a header block consists of
+the char ``type`` and uint64_t ``size``.
+Special objects start with the following type key:
+
+- ``c``: :class:`~qiskit.pulse.channels.Channel`
+- ``w``: :class:`~qiskit.pulse.library.Waveform`
+- ``s``: :class:`~qiskit.pulse.library.SymbolicPulse`
+
+.. _qpy_schedule_channel:
+
+CHANNEL
+-------
+
+Channel block starts with channel subtype ``char`` that maps an object data to
+:class:`~qiskit.pulse.channels.Channel` subclass. Mapping is defined as follows:
+
+- ``d``: :class:`~qiskit.pulse.channels.DriveChannel`
+- ``c``: :class:`~qiskit.pulse.channels.ControlChannel`
+- ``m``: :class:`~qiskit.pulse.channels.MeasureChannel`
+- ``a``: :class:`~qiskit.pulse.channels.AcquireChannel`
+- ``e``: :class:`~qiskit.pulse.channels.MemorySlot`
+- ``r``: :class:`~qiskit.pulse.channels.RegisterSlot`
+
+The key is immediately followed by the channel index serialized as the INSTRUCTION_PARAM.
+
+.. _qpy_schedule_waveform:
+
+Waveform
+--------
+
+Waveform block starts with WAVEFORM header:
+
+.. code-block:: c
+
+    struct {
+        double epsilon;
+        uint16_t data_size;
+        _bool amp_limited;
+    }
+
+which is followed by ``data_size`` bytes of numpy serialized complex ``ndarray`` in C-order,
```

Comment:
If we've moved to using npy here we should say that and rely on npy to handle the actual memory layout. I doubt people will be using fortran order manually in a waveform object but for the docs (and the code) I think it's better if we just abstract away the actual memory layout to numpy and not worry about it in the format specification.